runtime/near-vm-runner/src/wasmtime_runner.rs

```diff
@@ -0,0 +1,135 @@
+use crate::errors::IntoVMError;
+use crate::{imports, prepare};
+use near_runtime_fees::RuntimeFeesConfig;
+use near_vm_errors::FunctionCallError::LinkError;
+use near_vm_errors::{FunctionCallError, MethodResolveError, VMError, VMLogicError};
+use near_vm_logic::types::PromiseResult;
+use near_vm_logic::{External, MemoryLike, VMConfig, VMContext, VMLogic, VMOutcome};
+use std::ffi::c_void;
+use std::str;
+use wasmtime::{Engine, Limits, Linker, Memory, MemoryType, Module, Store};
+
+pub struct WasmtimeMemory(Memory);
+
+impl WasmtimeMemory {
+    pub fn new(
+        store: &Store,
+        initial_memory_bytes: u32,
+        max_memory_bytes: u32,
+    ) -> Result<Self, VMError> {
+        Ok(WasmtimeMemory(Memory::new(
+            store,
+            MemoryType::new(Limits::new(initial_memory_bytes, Some(max_memory_bytes))),
+        )))
+    }
+
+    pub fn clone(&self) -> Memory {
+        self.0.clone()
+    }
+}
+
+impl MemoryLike for WasmtimeMemory {
+    fn fits_memory(&self, offset: u64, len: u64) -> bool {
+        match offset.checked_add(len) {
+            None => false,
+            Some(end) => self.0.size() as u64 >= end,
+        }
+    }
+
+    fn read_memory(&self, offset: u64, buffer: &mut [u8]) {
+        let offset = offset as usize;
+        unsafe {
```

Comment:
Is it enforced by MemoryLike trait contract that we can use `unsafe` here? Can we have comments around every `unsafe` usage on why it is safe and minimize the block to the specific unsafe expression, so it is easier to see which method is exactly unsafe?

Reply:
Yeah, added comments.